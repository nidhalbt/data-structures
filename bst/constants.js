export const NIL_NODE = null;
export const NIL_VALUE = null;

export const isNil = (node) => {
  return !node;
};
