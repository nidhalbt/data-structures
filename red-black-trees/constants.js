export const COLORS = {
  RED: 'red',
  BLACK: 'black'
};
