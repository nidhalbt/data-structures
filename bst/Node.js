import { isNil, NIL_NODE } from "./constants";

export class Node {
  constructor(value, parent, left, right) {
    this.value = value;
    this.parent = parent;
    this.left = left;
    this.right = right;
    this.key = Math.random() * (1 - Math.random()) * Date.now();
  }

  newChild(value) {
    return new Node(value, this, NIL_NODE, NIL_NODE);
  }

  equals(node) {
    return !isNil(node) && (this.key === node.key);
  }

  push(value) {
    if (!this.value) {
      this.value = value;
      return this;
    }

    if (value < this.value) {
      return this.pushLeft(value);
    } else {
      return this.pushRight(value);
    }
  }

  pushLeft(value) {
    if (isNil(this.left)) {
      this.left = this.newChild(value);
      return this.left;
    }

    return this.left.push(value);
  }

  pushRight(value) {
    if (isNil(this.right)) {
      this.right = this.newChild(value);
      return this.right;
    }

    return this.right.push(value);
  }

  reverseLineage(child) {
    if (isNil(child)) {
      throw new Error('reverseLineage should NOT be called with an empty node');
    }

    if (isNil(child.parent) || !child.parent.equals(this)) {
      throw new Error('reverseLineage should not called on parent node to switch with its immediate child');
    }

    if (this.isLeftChild) {
      this.parent.left = child;
    }
    if (this.isRightChild) {
      this.parent.right = child;
    }

    child.parent = this.parent;
    this.parent = child;
  }

  rotateRight() {
    const { left } = this;
    this.reverseLineage(left);

    this.left = left.right;
    if (!isNil(this.left)) {
      this.left.parent = this;
    }
    left.right = this;
  }

  rotateLeft() {
    const { right } = this;
    this.reverseLineage(right);

    this.right = right.left;
    if (!isNil(this.right)) {
      this.right.parent = this;
    }
    right.left = this;
  }

  stripKeys() {
    delete this.key;
    if (!isNil(this.left)) {
      this.left.stripKeys();
    }
    if (!isNil(this.right)) {
      this.right.stripKeys();
    }
  }

  get isRoot() {
    return isNil(this.parent);
  }

  get isLeftChild() {
    return !isNil(this.parent) && this.equals(this.parent.left);
  }

  get isRightChild() {
    return !isNil(this.parent) && this.equals(this.parent.right);
  }

  get uncle() {
    const { isRoot, parent, grandparent } = this;
    if (isRoot || parent.isRoot) {
      return NIL_NODE;
    }

    return parent.isLeftChild ? grandparent.right : grandparent.left;
  }

  get grandparent() {
    if (isNil(this.parent) || isNil(this.parent.parent)) {
      return NIL_NODE;
    }

    return this.parent.parent;
  }
}
