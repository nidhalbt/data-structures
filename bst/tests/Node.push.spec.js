import { NIL_NODE } from "../constants";
import { Node } from "../Node";

describe('Node.push', () => {
  let root, left, right;

  beforeEach(() => {
    root = new Node(20, NIL_NODE, NIL_NODE, NIL_NODE);
    left = new Node(10, root, NIL_NODE, NIL_NODE);
    right = new Node(30, root, NIL_NODE, NIL_NODE);
    root.left = left;
    root.right = right;
  });

  it('should push be left.left', () => {
    root.push(5);
    expect(left.left).toBeTruthy();
    expect(left.left.value).toBe(5);
  });

  it('should push be left.right', () => {
    root.push(15);
    expect(left.right).toBeTruthy();
    expect(left.right.value).toBe(15);
  });

  it('should push be left.right', () => {
    root.push(10);
    expect(left.right).toBeTruthy();
    expect(left.right.value).toBe(10);
  });

  it('should push be right.left', () => {
    root.push(25);
    expect(right.left).toBeTruthy();
    expect(right.left.value).toBe(25);
  });

  it('should push be right.left', () => {
    root.push(20);
    expect(right.left).toBeTruthy();
    expect(right.left.value).toBe(20);
  });

  it('should push be right.right', () => {
    root.push(35);
    expect(right.right).toBeTruthy();
    expect(right.right.value).toBe(35);
  });

  it('various tests', () => {
    root.push(5);
    root.push(0);
    root.push(15);
    root.push(9);
    root.push(30);
    root.push(10);
    root.push(12);
    root.push(33);
    root.push(29);
    expect(left.left.value).toBe(5);
    expect(left.left.left.value).toBe(0);
    expect(left.left.right.value).toBe(9);
    expect(left.right.value).toBe(15);
    expect(left.right.left.value).toBe(10);
    expect(left.right.left.right.value).toBe(12);
    expect(right.left.value).toBe(29);
    expect(right.right.value).toBe(30);
    expect(right.right.right.value).toBe(33);
  });

  it('a snapshot test', () => {
    const values = [40, 15, 46, 30, 41, 15, 26, 15, 45, 8, 17, 24, 23, 49, 38, 39, 24, 7, 34, 32];
    values.forEach(e => root.push(e));
    root.stripKeys();
    expect(root).toMatchSnapshot();
  });
});
