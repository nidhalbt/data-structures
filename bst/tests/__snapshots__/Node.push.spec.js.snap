// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node.push a snapshot test 1`] = `
Node {
  "left": Node {
    "left": Node {
      "left": Node {
        "left": null,
        "parent": [Circular],
        "right": null,
        "value": 7,
      },
      "parent": [Circular],
      "right": null,
      "value": 8,
    },
    "parent": [Circular],
    "right": Node {
      "left": null,
      "parent": [Circular],
      "right": Node {
        "left": null,
        "parent": [Circular],
        "right": Node {
          "left": null,
          "parent": [Circular],
          "right": Node {
            "left": null,
            "parent": [Circular],
            "right": null,
            "value": 17,
          },
          "value": 15,
        },
        "value": 15,
      },
      "value": 15,
    },
    "value": 10,
  },
  "parent": null,
  "right": Node {
    "left": Node {
      "left": Node {
        "left": Node {
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 23,
        },
        "parent": [Circular],
        "right": Node {
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 24,
        },
        "value": 24,
      },
      "parent": [Circular],
      "right": null,
      "value": 26,
    },
    "parent": [Circular],
    "right": Node {
      "left": Node {
        "left": null,
        "parent": [Circular],
        "right": Node {
          "left": Node {
            "left": Node {
              "left": null,
              "parent": [Circular],
              "right": null,
              "value": 32,
            },
            "parent": [Circular],
            "right": null,
            "value": 34,
          },
          "parent": [Circular],
          "right": Node {
            "left": null,
            "parent": [Circular],
            "right": null,
            "value": 39,
          },
          "value": 38,
        },
        "value": 30,
      },
      "parent": [Circular],
      "right": Node {
        "left": Node {
          "left": null,
          "parent": [Circular],
          "right": Node {
            "left": null,
            "parent": [Circular],
            "right": null,
            "value": 45,
          },
          "value": 41,
        },
        "parent": [Circular],
        "right": Node {
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 49,
        },
        "value": 46,
      },
      "value": 40,
    },
    "value": 30,
  },
  "value": 20,
}
`;
